profiler/telemetry_test.go

```diff
@@ -0,0 +1,93 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package profiler
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/telemetry"
+)
+
+// Test that telemetry is enabled by default for the profiler. Turning on the profiler should send two telemetry events:
+// `app-product-change` to inform that the profiler has been enabled
+// `app-client-configuration-change` to send profiler-related configuration information
+func TestTelemetryEnabled(t *testing.T) {
+	t.Setenv("DD_TELEMETRY_HEARTBEAT_INTERVAL", "1")
+	receivedProducts := make(chan *telemetry.Products, 1)
+	configChanges := make(chan *telemetry.ConfigurationChange, 1)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/telemetry/proxy/api/v2/apmtelemetry" {
+			return
+		}
+		if r.Header.Get("DD-Telemetry-Request-Type") == string(telemetry.RequestTypeAppProductChange) {
+			var body telemetry.Body
+			body.Payload = new(telemetry.Products)
+			err := json.NewDecoder(r.Body).Decode(&body)
+			if err != nil {
+				t.Errorf("bad body: %s", err)
+			}
+			select {
+			case receivedProducts <- body.Payload.(*telemetry.Products):
+			default:
+			}
+		}
+		if r.Header.Get("DD-Telemetry-Request-Type") == string(telemetry.RequestTypeAppClientConfigurationChange) {
+			var body telemetry.Body
+			body.Payload = new(telemetry.ConfigurationChange)
+			err := json.NewDecoder(r.Body).Decode(&body)
+			if err != nil {
+				t.Errorf("bad body: %s", err)
+			}
+			select {
+			case configChanges <- body.Payload.(*telemetry.ConfigurationChange):
+			default:
+			}
+		}
+	}))
+	defer server.Close()
+	tracer.Start(tracer.WithAgentAddr(server.Listener.Addr().String()))
+	Start(
+		WithAgentAddr(server.Listener.Addr().String()),
+		WithProfileTypes(
+			BlockProfile,
+			HeapProfile,
+			MutexProfile,
+		),
+		WithPeriod(10*time.Millisecond),
+		CPUDuration(1*time.Millisecond),
+	)
+	defer Stop()
+
+	var productsPayload *telemetry.Products = <-receivedProducts
```

Comment:
I don't think an order is strictly guaranteed? The config change message is queued before the product change message in the `ProductChange` method, so I would expect the config change message to arrive first most of the time. 

I switched around the order and ran the test multiple times, and didn't seem to run into any problems.

Regardless with the order, wouldn't the buffered channels prevent the test from getting stuck? Since eventually both messages would arrive and be put into their respective buffers by the handler, and the receiving side can retrieve them from the channels in any order.

Reply:
Right, I missed that they were buffered 🤦 Yep, this is good then!